core/shared/src/main/scala/org/http4s/multipart/PartReceiver.scala

```diff
@@ -0,0 +1,206 @@
+import org.http4s.EntityDecoder
+import org.http4s.Headers
+import org.http4s.InvalidMessageBodyFailure
+
+/** Represents the decoding process of a single "part" in a `multipart/form-data` message.
+  *
+  * Used in conjunction with `MultipartReceiver` and `MultipartDecoder.fromReceiver`
+  * to create robust `EntityDecoder`s with fail-fast behavior.
+  */
+trait PartReceiver[F[_], A] {
+  def receive(part: Part[F]): Resource[F, Either[DecodeFailure, A]]
+
+  def map[B](f: A => B): PartReceiver[F, B] =
+    part => this.receive(part).map(_.map(f))
+
+  def mapWithHeaders[B](f: (Headers, A) => B): PartReceiver[F, B] =
+    part => this.receive(part).map(_.map(f(part.headers, _)))
+
+  def tapStart(onStart: F[Unit]): PartReceiver[F, A] =
+    part => Resource.eval(onStart).flatMap(_ => this.receive(part))
+
+  def tapResult(onResult: Either[DecodeFailure, A] => F[Unit]): PartReceiver[F, A] =
+    part => this.receive(part).evalTap(onResult)
+
+  def tapRelease(onRelease: F[Unit])(implicit F: Applicative[F]): PartReceiver[F, A] =
+    part => Resource.make(F.unit)(_ => onRelease).flatMap(_ => this.receive(part))
+
+  def preprocess(transformPartBody: Pipe[F, Byte, Byte]): PartReceiver[F, A] =
+    part => this.receive(part.copy(body = transformPartBody(part.body)))
+
+  def withSizeLimit(limit: Long)(implicit F: ApplicativeError[F, Throwable]): PartReceiver[F, A] =
+    preprocess(PartReceiver.limitPartSize[F](limit))
+
+  def rejectIfFile: PartReceiver[F, A] =
+    part =>
+      if (part.filename.isDefined) {
+        val effectivePartName = part.name.fold("<unnamed part")(name => s"'$name'")
+        Resource.pure(
+          Left(InvalidMessageBodyFailure(s"File data not allowed in $effectivePartName"))
+        )
+      } else {
+        this.receive(part)
+      }
+
+  def rejectIfNotFile: PartReceiver[F, A] =
+    part =>
+      if (part.filename.isEmpty) {
+        val effectivePartName = part.name.fold("<unnamed part")(name => s"'$name'")
+        Resource.pure(Left(InvalidMessageBodyFailure(s"File data required in $effectivePartName")))
+      } else {
+        this.receive(part)
+      }
+}
+
+object PartReceiver {
+
+  /** Creates a PartReceiver which decodes the part body to a String.
+    *
+    * The decoding will use UTF-8 unless the part provides a `Content-Type` header indicating otherwise.
+    */
+  def bodyText[F[_]](implicit F: Concurrent[F]): PartReceiver[F, String] =
+    part => Resource.eval(part.bodyText.compile.string).map(Right(_))
+
+  /** Creates a PartReceiver which writes the part body to a temporary file, then returns that file's `Path`. */
+  def toTempFile[F[_]](implicit F: Files[F], c: Compiler[F, F]): PartReceiver[F, Path] =
+    part =>
+      F.tempFile
+        .evalTap(path => part.body.through(F.writeAll(path)).compile.drain)
+        .map(Right(_))
+
+  /** Creates a PartReceiver that ignores the part body. */
+  def ignore[F[_]]: PartReceiver[F, Unit] =
+    _ => Resource.pure(Right(()))
```

Comment:
I think this is `Resource.unit`, or maybe `Applicative[Resource[F, *]].unit`?

Reply:
Unless that `*` is doing some heavy lifting, I think I'd still have to `.map(Right(_))` it, since the abstract method being fulfilled here is expected to return `Resource[F, Either[DecodeFailure, A]]`.